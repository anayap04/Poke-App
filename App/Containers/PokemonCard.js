import * as React from 'react';
import { Avatar, Button, Card, Title, Paragraph } from 'react-native-paper';



const PokemonCard = props => {
  const {
    pokemonName,
    pokemonImage,
    pokemonText,
    pokemonType,
  } = props;

  return (
  <Card>
    <Card.Title title="Card Title" subtitle="Card Subtitle" />
    <Card.Content>
      <Title>{pokemonName}</Title>
      <Paragraph>{pokemonText} {pokemonType}</Paragraph>
    </Card.Content>
    <Card.Cover source={{ uri: pokemonImage}} />
    <Card.Actions>
      <Button>Cancel</Button>
      <Button>Ok</Button>
    </Card.Actions>
  </Card>
  )
};

export default PokemonCard;