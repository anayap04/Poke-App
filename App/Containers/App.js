import React, {Component} from 'react';
import {connect} from 'react-redux';
import {SafeAreaView,View, Text, ScrollView, Image} from 'react-native';
import {Appbar, TextInput, Button} from 'react-native-paper';

//Components
import PokemonCard from './PokemonCard';

//Actions
import {
  requestAbilities,
  requestDesc,
  requestForm,
  requestPokedex,
  requestType,
} from '../Redux/Actions';

//Styles
import styles from './Styles/AppStyles';

//Utils
import {mapInfo} from '../Utils/pokedexUtils';

/**
 * App container
 */
export class App extends Component {
  constructor(props) {
    super(props);

    this.state = {
      pokemon: '',
    }
  }

  componentDidUpdate(prevProps) {
    const {pokedexData} = this.props;
    if (prevProps.pokedexData !== pokedexData) {
      this.getPokemonData(pokedexData);
    }
  }
  
  getPokemonData = (data) => {
    const {form, desc, type, abilities} = this.props;
    const urlForm = data.forms ? data.forms[0].url : null;
    const urlDesc =  data.species ? data.species.url : null;
    const urlType = data.types ? data.types[0].type.url : null;
    const arrayUrl = data.abilities 
      ? data.abilities.map(a => a.ability.url) 
      : null;
  
    form(urlForm);
    desc(urlDesc);
    type(urlType);
    abilities(arrayUrl);
  }

  /**
   * Una vez encuentra que hay cmabio en el estado,
   * esta función mappea el texto para evitar 
   * inyección de código y hace la llamada al servicio
   * @param {*} pokemon 
   */
  callService = pokemon => {
    const {request} = this.props;
    const newText = pokemon.replace(/[&\/\\#,+()$~%.'":*?<>{}]/g, '');    
    request(newText.toLowerCase());
    this.resetText();
  }

  /**
   * En esta función además de ingresar el pokemon a buscar,
   * @param {*} text ingresado por el usuario
   */
  setText = text => {
    this.setState({pokemon: text});
  }

  resetText = () => {
    this.setState({pokemon: ''})
  }

  /**
   * En esta función se renderiza el contenido mientras no haya error
   */
  renderContent = () => {
    const {
      formData,
      descData,
      typeData,
      abilitiesData,
    } = this.props;

    const pokemonInfo = 
      formData && 
      descData &&
      typeData &&
      abilitiesData
        ? mapInfo(
          formData,
          descData,
          typeData,
          abilitiesData) 
        : null;

    return (pokemonInfo ?
        <View>
          <PokemonCard 
            pokemonImage={pokemonInfo.pokemonImage}
            pokemonName={pokemonInfo.pokemonName}
            pokemonText={pokemonInfo.pokemonText}
            pokemonType={pokemonInfo.pokemonType}
            pokemonAbilities={pokemonInfo.pokemonAbilities}
          /> 
        </View> 
      :  <Image 
      style={styles.imageLogo}
      source={require('../../assets/images/descarga.png')}
    />
    );
  };

  renderError = () => {
    const {pokedexError} = this.props;
    return (
      <View style={styles.contentError}>
        <Text style={styles.title}>{pokedexError}</Text>
        <Text style={styles.subtitle}>
          {`El pokemon '${this.state.pokemon}' no aparece`}
        </Text>
      </View>)
  }

  render() {
    const {pokedexError} = this.props;

    return( 
    <View style={styles.container}>
      <Appbar style={styles.barStyle} />
        <SafeAreaView style={styles.safeAreaView}>
        <ScrollView style={styles.scrollView}>
        <TextInput
          label="Busca Tu pokemon"
          value={this.state.pokemon}
          onChangeText={text => this.setText(text)}
        />
        <Button 
          mode="contained" 
          style={styles.buttonStyle}
          onPress={() => this.callService(this.state.pokemon)}>
          <Text>Buscar</Text>
        </Button>
        {!pokedexError ? this.renderContent() : this.renderError()}
        </ScrollView>
      </SafeAreaView>
    </View>)
  }

}

const mapStateToProps = (state) => {
  const {
    pokedexData,
    pokedexError,
    formData,
    descData,
    typeData,
    abilitiesData,
  } = state;

  return {
    pokedexData: pokedexData,
    pokedexError: pokedexError,
    formData: formData,
    descData: descData,
    typeData: typeData,
    abilitiesData: abilitiesData,
  }
}

const mapDispatchToProps = dispatch => {
  return {
    request: pokemon => dispatch(requestPokedex(pokemon)),
    form: url => dispatch(requestForm(url)),
    desc: url => dispatch(requestDesc(url)),
    type: url => dispatch(requestType(url)),
    abilities: arrayUrl => dispatch(requestAbilities(arrayUrl)),
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(App)
