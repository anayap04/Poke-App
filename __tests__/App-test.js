import 'react-native';
import React from 'react';
import {App, mapStateToProps, mapDispatchToProps} from '../App/Containers/App';

// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';

const props = {
  pokedexData: {},
  pokedexError: null,
  formData: {},
  descData: {},
  typeData: {},
  abilitiesData: {},
};

it('renders correctly', () => {
  const component = renderer.create(<App {...props} />);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot(); 
});

it('mapToStateProps Test', () => {

  const state = {
    pokedexData: {},
    pokedexError: {},
    formData: {},
    descData: {},
    typeData: {},
    abilitiesData: {},
  };
  const result = mapStateToProps(state);
  expect(result.pokedexData).toEqual(state.pokedexData);
  expect(result.pokedexError).toEqual(state.pokedexError);
  expect(result.formData).toEqual(state.formData);
  expect(result.descData).toEqual(state.descData);
  expect(result.typeData).toEqual(state.typeData);
  expect(result.abilitiesData).toEqual(state.abilitiesData);
})
