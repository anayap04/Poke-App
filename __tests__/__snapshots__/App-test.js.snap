// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#CC0000",
      "flex": 1,
      "height": "100%",
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#CC0000",
        "bottom": 0,
        "elevation": 4,
        "flexDirection": "row",
        "height": 56,
        "left": 0,
        "paddingHorizontal": 4,
        "right": 0,
        "shadowColor": "#000000",
        "shadowOffset": Object {
          "height": 3,
          "width": 0,
        },
        "shadowOpacity": 0.24,
        "shadowRadius": 4,
      }
    }
  />
  <RCTSafeAreaView
    emulateUnlessSupported={true}
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <RCTScrollView
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <View>
        <View
          style={
            Array [
              Object {
                "backgroundColor": "rgb(231, 231, 231)",
                "borderTopLeftRadius": 4,
                "borderTopRightRadius": 4,
              },
              Object {},
            ]
          }
        >
          <View
            style={
              Object {
                "backgroundColor": "rgba(0, 0, 0, 0.26)",
                "bottom": 0,
                "height": 2,
                "left": 0,
                "position": "absolute",
                "right": 0,
                "transform": Array [
                  Object {
                    "scaleY": 0.5,
                  },
                ],
              }
            }
          />
          <View
            style={
              Array [
                Object {
                  "paddingBottom": 0,
                  "paddingTop": 0,
                },
                Object {
                  "minHeight": 64,
                },
              ]
            }
          >
            <View
              pointerEvents="none"
              style={
                Object {
                  "bottom": 0,
                  "left": 0,
                  "opacity": 1,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                  "transform": Array [
                    Object {
                      "translateX": 0,
                    },
                  ],
                }
              }
            >
              <Text
                numberOfLines={1}
                onLayout={[Function]}
                style={
                  Object {
                    "color": "#6200ee",
                    "fontFamily": "System",
                    "fontSize": 16,
                    "fontWeight": undefined,
                    "left": 0,
                    "opacity": 0,
                    "paddingLeft": 12,
                    "paddingRight": 12,
                    "position": "absolute",
                    "textAlign": "left",
                    "top": 34,
                    "transform": Array [
                      Object {
                        "translateX": 0,
                      },
                      Object {
                        "translateY": 0,
                      },
                      Object {
                        "scale": 1,
                      },
                    ],
                    "writingDirection": "ltr",
                  }
                }
              >
                Busca Tu pokemon
              </Text>
              <Text
                numberOfLines={1}
                style={
                  Object {
                    "color": "rgba(0, 0, 0, 0.54)",
                    "fontFamily": "System",
                    "fontSize": 16,
                    "fontWeight": undefined,
                    "left": 0,
                    "opacity": 0,
                    "paddingLeft": 12,
                    "paddingRight": 12,
                    "position": "absolute",
                    "textAlign": "left",
                    "top": 34,
                    "transform": Array [
                      Object {
                        "translateX": 0,
                      },
                      Object {
                        "translateY": 0,
                      },
                      Object {
                        "scale": 1,
                      },
                    ],
                    "writingDirection": "ltr",
                  }
                }
              >
                Busca Tu pokemon
              </Text>
            </View>
            <TextInput
              allowFontScaling={true}
              editable={true}
              multiline={false}
              onBlur={[Function]}
              onChangeText={[Function]}
              onFocus={[Function]}
              placeholder=""
              placeholderTextColor="rgba(0, 0, 0, 0.54)"
              rejectResponderTermination={true}
              selectionColor="#6200ee"
              style={
                Array [
                  Object {
                    "flexGrow": 1,
                    "margin": 0,
                    "zIndex": 1,
                  },
                  Object {
                    "paddingLeft": 12,
                    "paddingRight": 12,
                  },
                  Object {
                    "height": 64,
                  },
                  Object {
                    "paddingBottom": 4,
                    "paddingTop": 24,
                  },
                  Object {
                    "color": "#000000",
                    "fontFamily": "System",
                    "fontSize": 16,
                    "fontWeight": undefined,
                    "textAlign": "left",
                    "textAlignVertical": "center",
                  },
                  Array [
                    Object {},
                  ],
                ]
              }
              underlineColorAndroid="transparent"
              value=""
            />
          </View>
        </View>
        <View
          style={
            Object {
              "backgroundColor": "#3B4CCA",
              "borderColor": "white",
              "borderRadius": 9,
              "borderStyle": "solid",
              "borderWidth": 1,
              "elevation": 12,
              "marginTop": 10,
              "minWidth": 64,
              "shadowColor": "#000",
              "shadowOffset": Object {
                "height": 6,
                "width": 0,
              },
              "shadowOpacity": 0.37,
              "shadowRadius": 7.49,
            }
          }
        >
          <View
            accessibilityRole="button"
            accessibilityState={
              Object {
                "disabled": undefined,
              }
            }
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Array [
                Object {
                  "overflow": "hidden",
                },
                Object {
                  "borderRadius": 9,
                },
              ]
            }
          >
            <View
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "justifyContent": "center",
                  },
                  undefined,
                ]
              }
            >
              <Text
                numberOfLines={1}
                selectable={false}
                style={
                  Array [
                    Object {
                      "color": "#000000",
                      "fontFamily": "System",
                      "fontWeight": "400",
                    },
                    Object {
                      "textAlign": "left",
                    },
                    Array [
                      Object {
                        "letterSpacing": 1,
                        "marginHorizontal": 16,
                        "marginVertical": 9,
                        "textAlign": "center",
                      },
                      undefined,
                      Object {
                        "textTransform": "uppercase",
                      },
                      Object {
                        "color": "#ffffff",
                        "fontFamily": "System",
                        "fontWeight": "500",
                      },
                      Object {
                        "fontFamily": "System",
                        "fontWeight": "500",
                      },
                      undefined,
                    ],
                  ]
                }
              >
                <Text>
                  Buscar
                </Text>
              </Text>
            </View>
          </View>
        </View>
        <View>
          <View>
            <View
              style={
                Object {
                  "alignContent": "center",
                  "alignItems": "center",
                  "paddingBottom": 20,
                  "paddingTop": 20,
                }
              }
            >
              <Text
                style={
                  Array [
                    Object {
                      "color": "#000000",
                      "fontFamily": "System",
                      "fontWeight": "400",
                    },
                    Object {
                      "textAlign": "left",
                    },
                    Array [
                      Object {
                        "textAlign": "left",
                      },
                      Object {
                        "color": "rgba(0, 0, 0, 0.87)",
                        "fontFamily": "System",
                        "fontWeight": "500",
                        "writingDirection": "ltr",
                      },
                      Array [
                        Object {
                          "fontSize": 20,
                          "letterSpacing": 0.15,
                          "lineHeight": 30,
                          "marginVertical": 2,
                        },
                        Object {
                          "color": "#FAFAFA",
                          "fontFamily": "Lato-Black",
                          "fontSize": 32,
                        },
                      ],
                    ],
                  ]
                }
              />
              <View
                style={
                  Object {
                    "borderColor": "#777",
                    "borderRadius": 6,
                    "borderWidth": 4,
                    "paddingLeft": 4,
                    "paddingRight": 4,
                  }
                }
              >
                <Text
                  style={
                    Array [
                      Object {
                        "color": "#000000",
                        "fontFamily": "System",
                        "fontWeight": "400",
                      },
                      Object {
                        "textAlign": "left",
                      },
                      Array [
                        Object {
                          "textAlign": "left",
                        },
                        Object {
                          "color": "rgba(0, 0, 0, 0.87)",
                          "fontFamily": "System",
                          "fontWeight": "400",
                          "writingDirection": "ltr",
                        },
                        Array [
                          Object {
                            "fontSize": 16,
                            "letterSpacing": 0.5,
                            "lineHeight": 24,
                            "marginVertical": 2,
                          },
                          Array [
                            Object {
                              "color": "#777",
                              "fontFamily": "Lato-Blod",
                              "fontSize": 20,
                            },
                          ],
                        ],
                      ],
                    ]
                  }
                />
              </View>
            </View>
            <View
              style={
                Object {
                  "backgroundColor": "#777",
                  "borderColor": "#DEE4E6",
                  "borderRadius": 2,
                  "borderWidth": 20,
                }
              }
            >
              <Image
                source={
                  Object {
                    "uri": null,
                  }
                }
                style={
                  Object {
                    "height": 250,
                    "width": "100%",
                  }
                }
              />
            </View>
            <View
              style={
                Object {
                  "backgroundColor": "#CC0000",
                  "flexDirection": "row",
                  "paddingBottom": 10,
                  "paddingLeft": 4,
                  "paddingRight": 4,
                  "paddingTop": 10,
                }
              }
            >
              <View
                style={
                  Object {
                    "alignContent": "center",
                    "backgroundColor": "#3B4CCA",
                    "borderColor": "white",
                    "borderRadius": 9,
                    "borderStyle": "solid",
                    "borderWidth": 1,
                    "elevation": 12,
                    "flexDirection": "row",
                    "justifyContent": "center",
                    "marginLeft": 10,
                    "shadowColor": "#000",
                    "shadowOffset": Object {
                      "height": 6,
                      "width": 0,
                    },
                    "shadowOpacity": 0.37,
                    "shadowRadius": 7.49,
                    "width": "45%",
                  }
                }
              >
                <View
                  accessibilityRole="button"
                  accessibilityState={
                    Object {
                      "disabled": false,
                      "selected": false,
                    }
                  }
                  accessible={true}
                  focusable={true}
                  onClick={[Function]}
                  onResponderGrant={[Function]}
                  onResponderMove={[Function]}
                  onResponderRelease={[Function]}
                  onResponderTerminate={[Function]}
                  onResponderTerminationRequest={[Function]}
                  onStartShouldSetResponder={[Function]}
                  style={
                    Array [
                      Object {
                        "overflow": "hidden",
                      },
                      Object {
                        "borderRadius": 9,
                      },
                    ]
                  }
                >
                  <View
                    style={
                      Array [
                        Object {
                          "alignItems": "center",
                          "flexDirection": "row",
                          "paddingLeft": 4,
                          "position": "relative",
                        },
                        Object {
                          "paddingRight": 4,
                        },
                      ]
                    }
                  >
                    <Text
                      numberOfLines={1}
                      selectable={false}
                      style={
                        Array [
                          Object {
                            "color": "#000000",
                            "fontFamily": "System",
                            "fontWeight": "400",
                          },
                          Object {
                            "textAlign": "left",
                          },
                          Array [
                            Object {
                              "lineHeight": 24,
                              "marginVertical": 4,
                              "minHeight": 24,
                              "textAlignVertical": "center",
                            },
                            Object {
                              "color": "rgba(0, 0, 0, 0.87)",
                              "fontFamily": "System",
                              "fontWeight": "400",
                              "marginLeft": 8,
                              "marginRight": 8,
                            },
                            undefined,
                          ],
                        ]
                      }
                    >
                      <Text
                        style={
                          Array [
                            Object {
                              "color": "#000000",
                              "fontFamily": "System",
                              "fontWeight": "400",
                            },
                            Object {
                              "textAlign": "left",
                            },
                            Object {
                              "color": "white",
                              "fontFamily": "Lato-Blod",
                            },
                          ]
                        }
                      >
                        Información
                      </Text>
                    </Text>
                  </View>
                </View>
              </View>
              <View
                style={
                  Object {
                    "alignContent": "center",
                    "backgroundColor": "#3B4CCA",
                    "borderColor": "white",
                    "borderRadius": 9,
                    "borderStyle": "solid",
                    "borderWidth": 1,
                    "elevation": 12,
                    "flexDirection": "row",
                    "justifyContent": "center",
                    "marginLeft": 10,
                    "shadowColor": "#000",
                    "shadowOffset": Object {
                      "height": 6,
                      "width": 0,
                    },
                    "shadowOpacity": 0.37,
                    "shadowRadius": 7.49,
                    "width": "45%",
                  }
                }
              >
                <View
                  accessibilityRole="button"
                  accessibilityState={
                    Object {
                      "disabled": false,
                      "selected": false,
                    }
                  }
                  accessible={true}
                  focusable={true}
                  onClick={[Function]}
                  onResponderGrant={[Function]}
                  onResponderMove={[Function]}
                  onResponderRelease={[Function]}
                  onResponderTerminate={[Function]}
                  onResponderTerminationRequest={[Function]}
                  onStartShouldSetResponder={[Function]}
                  style={
                    Array [
                      Object {
                        "overflow": "hidden",
                      },
                      Object {
                        "borderRadius": 9,
                      },
                    ]
                  }
                >
                  <View
                    style={
                      Array [
                        Object {
                          "alignItems": "center",
                          "flexDirection": "row",
                          "paddingLeft": 4,
                          "position": "relative",
                        },
                        Object {
                          "paddingRight": 4,
                        },
                      ]
                    }
                  >
                    <Text
                      numberOfLines={1}
                      selectable={false}
                      style={
                        Array [
                          Object {
                            "color": "#000000",
                            "fontFamily": "System",
                            "fontWeight": "400",
                          },
                          Object {
                            "textAlign": "left",
                          },
                          Array [
                            Object {
                              "lineHeight": 24,
                              "marginVertical": 4,
                              "minHeight": 24,
                              "textAlignVertical": "center",
                            },
                            Object {
                              "color": "rgba(0, 0, 0, 0.87)",
                              "fontFamily": "System",
                              "fontWeight": "400",
                              "marginLeft": 8,
                              "marginRight": 8,
                            },
                            undefined,
                          ],
                        ]
                      }
                    >
                      <Text
                        style={
                          Array [
                            Object {
                              "color": "#000000",
                              "fontFamily": "System",
                              "fontWeight": "400",
                            },
                            Object {
                              "textAlign": "left",
                            },
                            Object {
                              "color": "white",
                              "fontFamily": "Lato-Blod",
                            },
                          ]
                        }
                      >
                        Estadísticas
                      </Text>
                    </Text>
                  </View>
                </View>
              </View>
            </View>
            <View
              style={
                Object {
                  "alignSelf": "center",
                  "backgroundColor": "rgba(242, 234, 234, 0.8)",
                  "marginBottom": 10,
                  "marginTop": 10,
                  "padding": 20,
                  "width": "90%",
                }
              }
            >
              <Text
                style={
                  Array [
                    Object {
                      "color": "#000000",
                      "fontFamily": "System",
                      "fontWeight": "400",
                    },
                    Object {
                      "textAlign": "left",
                    },
                    Object {
                      "fontFamily": "Lato-Regular",
                      "fontSize": 20,
                    },
                  ]
                }
              />
            </View>
          </View>
        </View>
      </View>
    </RCTScrollView>
  </RCTSafeAreaView>
</View>
`;
